### PR TITLE
test: 동시성 문제로 발생하는 예외 처리 확인을 위한 테스트 작성 완료

### PR DESCRIPTION
**Pull request by yeonise**

### CODE REVIEW

**Comment by yeonise** [comment]
### Test Coverage Report
|Overall Project|91.81% **`-0.35%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|94.74%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[MemberRegisterException.java](https://github.com/mujik-tigers/dowith/blob/e26961d93c13d7d52a16f47d67fbef1f419e1249/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fexception%2Fmember%2FMemberRegisterException.java)|100%|:green_apple:|
|[AuthService.java](https://github.com/mujik-tigers/dowith/blob/e26961d93c13d7d52a16f47d67fbef1f419e1249/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fauth%2Fapplication%2FAuthService.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/mujik-tigers/dowith/blob/e26961d93c13d7d52a16f47d67fbef1f419e1249/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fmember%2Fapplication%2FMemberService.java)|92.31% **`-7.69%`**|:green_apple:|
|[LoginData.java](https://github.com/mujik-tigers/dowith/blob/e26961d93c13d7d52a16f47d67fbef1f419e1249/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fauth%2Fdto%2FLoginData.java)|88.89%|:green_apple:|
|[GlobalExceptionHandler.java](https://github.com/mujik-tigers/dowith/blob/e26961d93c13d7d52a16f47d67fbef1f419e1249/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fexception%2FGlobalExceptionHandler.java)|52.22%|:green_apple:|

File: src/test/java/com/dowe/auth/application/AuthServiceTest.java [hunk 1]
```diff
@@ -82,26 +87,45 @@ void testIsFirstTime2() {
...
+
+		for (int i = 0; i < threadCount; i++) {
+			executorService.submit(() -> {
+				try {
+					authService.login(provider, authorizationCode);
+					successCount.incrementAndGet();
+				} catch (MemberRegisterException exception) {
+					failureCount.incrementAndGet();
+				} finally {
+					latch.countDown();
```
Comment (ghkdgus29) [inline src/test/java/com/dowe/auth/application/AuthServiceTest.java:121]: 천재냥.

File: src/test/java/com/dowe/exception/GlobalExceptionHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+						.andExpect(status().isOk());
+
+					String response = resultActions.andReturn().getResponse().getContentAsString(Charsets.UTF_8);
+					JsonNode rootNode = objectMapper.readTree(response);
+					LoginData loginData = objectMapper.treeToValue(rootNode.get("data"), LoginData.class);
+
+					if (loginData.isFirstTime()) {
+						firstTimeTrueCount.incrementAndGet();
+					} else {
+						firstTimeFalseCount.incrementAndGet();
```
Comment (ghkdgus29) [inline src/test/java/com/dowe/exception/GlobalExceptionHandlerTest.java:93]: AtomicInteger를 이용하여 동시성 테스트를 하는거 정말 좋은 아이디어네요 굿굿!!

File: src/main/java/com/dowe/member/application/MemberService.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+import lombok.RequiredArgsConstructor;
+
+@Service
+@RequiredArgsConstructor
+public class MemberService {
+
+	private final MemberRepository memberRepository;
+
+	public Optional<Member> findByProvider(Provider provider, String authId) {
```
Comment (ghkdgus29) [inline src/main/java/com/dowe/member/application/MemberService.java]: `findMemberBy` 라는 메서드 명은 어떤가요?

**Comment by yeonise** [comment]
### Test Coverage Report
|Overall Project|92.29% **`-0.33%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|97.3%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[RedisConfig.java](https://github.com/mujik-tigers/dowith/blob/9f205940bf1be0e0e56d2ea485e2b8ece5718141/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fconfig%2FRedisConfig.java)|100%|:green_apple:|
|[MemberRegisterException.java](https://github.com/mujik-tigers/dowith/blob/9f205940bf1be0e0e56d2ea485e2b8ece5718141/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fexception%2Fmember%2FMemberRegisterException.java)|100%|:green_apple:|
|[AuthService.java](https://github.com/mujik-tigers/dowith/blob/9f205940bf1be0e0e56d2ea485e2b8ece5718141/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fauth%2Fapplication%2FAuthService.java)|100%|:green_apple:|
|[MemberCodeStorage.java](https://github.com/mujik-tigers/dowith/blob/9f205940bf1be0e0e56d2ea485e2b8ece5718141/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fmember%2Fapplication%2FMemberCodeStorage.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/mujik-tigers/dowith/blob/9f205940bf1be0e0e56d2ea485e2b8ece5718141/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fmember%2Fapplication%2FMemberService.java)|92.86% **`-7.14%`**|:green_apple:|
|[LoginData.java](https://github.com/mujik-tigers/dowith/blob/9f205940bf1be0e0e56d2ea485e2b8ece5718141/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fauth%2Fdto%2FLoginData.java)|88.89%|:green_apple:|
|[GlobalExceptionHandler.java](https://github.com/mujik-tigers/dowith/blob/9f205940bf1be0e0e56d2ea485e2b8ece5718141/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fexception%2FGlobalExceptionHandler.java)|52.22%|:green_apple:|

File: src/test/java/com/dowe/auth/application/TokenManagerTest.java [hunk 4]
```diff
@@ -32,7 +32,7 @@ class TokenManagerTest extends IntegrationTestSupport {
 	@Autowired
 	private JwtProperties jwtProperties;
 
-	@AfterEach
+	@BeforeEach
 	void clear() {
```
Comment (yeonise) [inline src/test/java/com/dowe/auth/application/TokenManagerTest.java:36]: 질문: After에서 Before로 변경하게 된 이유가 무었인가요?

File: src/main/resources/schema.sql [hunk 5]
```diff
@@ -6,7 +6,7 @@ CREATE TABLE `member` (
   `provider` varchar(10),
   `auth_id` varchar(255),
   `name` varchar(25),
-  `code` varchar(10) UNIQUE
+  `code` varchar(10)
```
Comment (yeonise) [inline src/main/resources/schema.sql:9]: 질문: 유니크 제약 조건을 제거한 이유가 무엇인가요?

만약 제약 조건을 지운다면,
MySQL에 이미 데이터가 존재하고 Redis를 추후에 띄운 경우에 대비하여
MySQL에 있는 code 데이터들을 모두 Redis에 밀어 넣어주는 작업이 필요할 것 같아요.

File: src/main/java/com/dowe/member/application/MemberService.java [hunk 6]
```diff
@@ -19,8 +19,9 @@
 public class MemberService {
 
 	private final MemberRepository memberRepository;
+	private final MemberCodeStorage memberCodeStorage;
 
-	public Optional<Member> findByProvider(Provider provider, String authId) {
+	public Optional<Member> findByProvider(Provider provider, String authId) {        // TODO: findMemberBy 메서드 명 변경 ㄱㅊ? -> yeonise의 허락
```
Comment (yeonise) [inline src/main/java/com/dowe/member/application/MemberService.java]: 의견: 저는 개인적으로 Member를 반환하는 MemberRepository이기 때문에 메서드에 Member를 생략하는 것이 좋다고 생각해요!
( `findById`, `findAll` 처럼 )

**Review by yeonise — COMMENTED** [review]
Redis의 특징과 lua script를 잘 활용한 사례 같아요!
수고하셨습니다 👏🏻

File: src/test/java/com/dowe/auth/application/TokenManagerTest.java (hunk 4 above)
Comment (ghkdgus29) [reply src/test/java/com/dowe/auth/application/TokenManagerTest.java:36]: Redis를 다른 테스트에서 clean하지 않는 경우, Redis를 사용하는 테스트의 값이 깨질 수 있어서 @Before로 바꿨습니당

File: src/main/resources/schema.sql (hunk 5 above)
Comment (ghkdgus29) [reply src/main/resources/schema.sql:9]: unique한 값이 무조건 레디스로부터 받아오기 때문에 unique 조건을 없앴어요.
말씀하신대로 Redis에 밀어 넣어주는 작업이 필요하겠네요.

File: src/main/java/com/dowe/member/application/MemberService.java (hunk 6 above)
Comment (ghkdgus29) [reply src/main/java/com/dowe/member/application/MemberService.java]: 그렇다면 findBy만 하는건 어때요?

Comment (yeonise) [reply src/main/java/com/dowe/member/application/MemberService.java]: 좋습니다~

**Comment by yeonise** [comment]
### Test Coverage Report
|Overall Project|89.69% **`-3.14%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|78.69%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[RedisConfig.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fconfig%2FRedisConfig.java)|100%|:green_apple:|
|[MemberRegisterException.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fexception%2Fmember%2FMemberRegisterException.java)|100%|:green_apple:|
|[AuthService.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fauth%2Fapplication%2FAuthService.java)|100%|:green_apple:|
|[MemberCodeStorage.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fmember%2Fapplication%2FMemberCodeStorage.java)|100%|:green_apple:|
|[MemberService.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fmember%2Fapplication%2FMemberService.java)|92.86% **`-7.14%`**|:green_apple:|
|[LoginData.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fauth%2Fdto%2FLoginData.java)|88.89%|:green_apple:|
|[GlobalExceptionHandler.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Fexception%2FGlobalExceptionHandler.java)|52.22%|:green_apple:|
|[Initializer.java](https://github.com/mujik-tigers/dowith/blob/e24ba6418cef34bded7c7b54b4ff0ef4681b22e6/src%2Fmain%2Fjava%2Fcom%2Fdowe%2Futil%2FInitializer.java)|0%|:x:|

**Pull request merged**